pmd-apex/src/main/resources/category/apex/design.xml

```diff
@@ -36,6 +36,44 @@ public class Foo {
         </example>
     </rule>
 
+    <rule name="AvoidUnusedMethod"
+          since="7.0.0"
+          message="Unused methods make understanding code harder"
+          class="net.sourceforge.pmd.lang.apex.rule.design.AvoidUnusedMethodRule"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_apex_design.html#avoidunusedmethod">
+        <description>
+Avoid having unused methods since they make understanding and maintaining code harder.
+
+This rule finds not only unused private methods, but public methods as well.
+
+[ApexLink](https://github.com/nawforce/ApexLink) is used to make this possible and this needs
+additional configuration. The environment variable `PMD_APEX_ROOTDIRECTORY` needs to be set prior to executing
+PMD. With this variable the root directory of the Salesforce metadata, where `sfdx-project.json` resides, is
+specified. ApexLink can then load all the classes in the project and figure out, whether a method is used or not.
+        </description>
```

Comment:
@nawforce Hi, as a last step before merging this, could you document as thoroughly as possible what the project structure needs to look like for this to work properly? From [this thread](https://github.com/pmd/pmd/pull/2830#discussion_r507057206) I could gather, that  
- The project needs to use the [SDFX format](https://developer.salesforce.com/docs/atlas.en-us.sfdx_dev.meta/sfdx_dev/sfdx_dev_source_file_format.htm) (would be nice to include this link)
- The `-dir` option needs to point to a subset of the source tree listed in the `sdfx-project.json`, otherwise we would just be pointing to some unrelated sources, but we don't actually check this and treat this as an error (yet)
- Given a source tree in the correct format and a `sdfx-project.json`, ApexLink will generate the missing metadata if need be, and everything should run smoothly in one step (no manual pre-generation or so).

Is this correct? 

I think the goal this doc should achieve, is that even an unexperienced developer, who doesn't know ApexLink, can get a pretty good idea of how to configure the rule, and what ApexLink will do to their files. I don't feel like I know enough about your tool or Apex in general to write this up myself.

Reply:
Sure, I will take a look at it tomorrow. Your description on this is pretty accurate for the common case but we may want to think about directing them to a page for more complex cases. A bit of feedback I have been getting is that when devs are using 2GP to modularise an application they have more complex needs than I am really allowing for currently so we need to be clear about this.

Reply:
@oowekyala I was just taking a look at latest on the branch and looks like PMD_APEX_ROOTDIRECTORY is not getting picked up as a property. It runs through the code but the getProperty(MULTIFILE_DIRECTORY) defaults to "" which results in nothing being found. Don't think I understand enough about how that should have worked to be able to spot where it's failing.

Reply:
I think c314ba8 should fix the problem. I can add a test tomorrow

Reply:
That's fixed it for me, thanks very much.

Reply:
I have extended the description with a couple of extra sections. The first just gives some background on what is needed in PMD_APEX_ROOTDIRECTORY for things to work, while the second describes how to workaround errors you might get if building over other packages. 

I added a reference in to turning on debug logging, not sure that is really what is needed here. What usually happens when the config is  wrong is that you get false positive errors reported against the source files but we are ignoring all but the Unused diagnostics at the moment. Maybe it would be better to direct any errors recorded against source files to the debug log automatically?

Reply:
> I have extended the description with a couple of extra sections. The first just gives some background on what is needed in PMD_APEX_ROOTDIRECTORY for things to work, while the second describes how to workaround errors you might get if building over other packages.

Thanks, this looks great! I managed setting it up after stumbling on https://github.com/forcedotcom/sfdx-simple. The source format page linked doesn't show any examples of what the project file should look like unfortunately. (but, I'm an apex noob)

> I added a reference in to turning on debug logging, not sure that is really what is needed here. What usually happens when the config is wrong is that you get false positive errors reported against the source files but we are ignoring all but the Unused diagnostics at the moment. Maybe it would be better to direct any errors recorded against source files to the debug log automatically?

What kind of false positive errors? Are they exceptions? In which case I think they should be logged as `severe`, so that the user knows what happens.

Now from what I've seen while trying it out, the errors are duplicated, one for each file, and the SEVERE log channel would get flooded... I'll push a fix

Reply:
Typically the errors are for missing types due to not setting up sfdx-project.json to include everything needed. There are a couple of reasons this happens with Salesforce projects more than it should but I will save that for another day. Logically I am thinking in a multi-file analysis these types of errors are quite similar to the current handling of say a Java syntax error and some may even be syntax errors.

It looks like that would normally be handled by throwing a ParseException which is then caught by the SourceCodeProcessor. Throwing multiple errors this way sounds like a bad idea but I am wondering if we log the errors and then throw a summary exception indicating problems were found might that be the right choice?

The API I have at the moment is not very useful for this in that it produces a String or JSON output of all errors/warnings in the Org. Would be better for me to add one to just grab the first 'n' errors so if config is badly wrong we won't spam them too much.

Reply:
Ok I see

> Throwing multiple errors this way sounds like a bad idea but I am wondering if we log the errors and then throw a summary exception indicating problems were found might that be the right choice?

This sounds good. Actually pmd 7 will have a concept of "semantic errors", which are reported during Parser.parse, but may be recovered from. For now the code is only in pmd-java and has not yet made it into pmd core, but [here's a draft of the API](https://github.com/oowekyala/pmd/blob/25425b1c804964d94de26fd022f19ca32743d135/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/SemanticErrorReporter.java#L15-L39). So we can report these as errors and still continue semantic analysis on the file, to collect more errors if needed, and report them in bulk later. And currently my stance is that semantic errors would prevent rules from being run later.

For now I would keep it as is and add a `fixme` comment to change the apexlink logger implementation that's in ApexMultifileAnalysis to forward to a SemanticErrorReporter when that's moved to pmd core.

Reply:
Ok, I will sort the fixme and see if I can resolve the other couple of issues outstanding

Reply:
Added FIXME to comment on this.